FTBFS: error: reference to ‘array’ is ambiguous

Tried to compile the latest tip of `devel` (2f474a804a381915770981f518493204bb53d3c1) on Debian, but getting a `error: reference to ‘array’ is ambiguous` in `test/clamp.cpp`. Sounds like someone perhaps brought `std::array` to the `af` namespace, hence the ambiguity?

Detailed error:
```
[ 70%] Building CXX object test/CMakeFiles/clamp_opencl.dir/clamp.cpp.o
/home/gvaillan/local/src/arrayfire/test/clamp.cpp: In member function ‘virtual void ClampTests_FloatArrayArray_Test::TestBody()’:
/home/gvaillan/local/src/arrayfire/test/clamp.cpp:24:5: error: reference to ‘array’ is ambiguous
     array in = af::randu(num, f32);
     ^~~~~
In file included from /home/gvaillan/local/src/arrayfire/test/clamp.cpp:11:0:
/home/gvaillan/local/src/arrayfire/include/af/array.h:27:17: note: candidates are: class af::array
     class AFAPI array {
                 ^~~~~
In file included from /usr/include/c++/6/tuple:39:0,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/internal/gtest-port.h:708,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/internal/gtest-internal.h:40,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/gtest.h:58,
                 from /home/gvaillan/local/src/arrayfire/test/clamp.cpp:10:
/usr/include/c++/6/array:90:12: note:                 template<class _Tp, long unsigned int _Nm> struct std::array
     struct array
            ^~~~~
/home/gvaillan/local/src/arrayfire/test/clamp.cpp:25:5: error: reference to ‘array’ is ambiguous
     array lo = af::randu(num, f32)/10;       // Ensure lo <= 0.1
     ^~~~~
In file included from /home/gvaillan/local/src/arrayfire/test/clamp.cpp:11:0:
/home/gvaillan/local/src/arrayfire/include/af/array.h:27:17: note: candidates are: class af::array
     class AFAPI array {
                 ^~~~~
In file included from /usr/include/c++/6/tuple:39:0,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/internal/gtest-port.h:708,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/internal/gtest-internal.h:40,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/gtest.h:58,
                 from /home/gvaillan/local/src/arrayfire/test/clamp.cpp:10:
/usr/include/c++/6/array:90:12: note:                 template<class _Tp, long unsigned int _Nm> struct std::array
     struct array
            ^~~~~
/home/gvaillan/local/src/arrayfire/test/clamp.cpp:26:5: error: reference to ‘array’ is ambiguous
     array hi = 1.0 - af::randu(num, f32)/10; // Ensure hi >= 0.9
     ^~~~~
In file included from /home/gvaillan/local/src/arrayfire/test/clamp.cpp:11:0:
/home/gvaillan/local/src/arrayfire/include/af/array.h:27:17: note: candidates are: class af::array
     class AFAPI array {
                 ^~~~~
In file included from /usr/include/c++/6/tuple:39:0,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/internal/gtest-port.h:708,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/internal/gtest-internal.h:40,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/gtest.h:58,
                 from /home/gvaillan/local/src/arrayfire/test/clamp.cpp:10:
/usr/include/c++/6/array:90:12: note:                 template<class _Tp, long unsigned int _Nm> struct std::array
     struct array
            ^~~~~
/home/gvaillan/local/src/arrayfire/test/clamp.cpp:27:14: error: ‘lo’ was not declared in this scope
     af::eval(lo, hi);
              ^~
/home/gvaillan/local/src/arrayfire/test/clamp.cpp:27:18: error: ‘hi’ was not declared in this scope
     af::eval(lo, hi);
                  ^~
/home/gvaillan/local/src/arrayfire/test/clamp.cpp:31:5: error: reference to ‘array’ is ambiguous
     array out = clamp(in, lo, hi);
     ^~~~~
In file included from /home/gvaillan/local/src/arrayfire/test/clamp.cpp:11:0:
/home/gvaillan/local/src/arrayfire/include/af/array.h:27:17: note: candidates are: class af::array
     class AFAPI array {
                 ^~~~~
In file included from /usr/include/c++/6/tuple:39:0,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/internal/gtest-port.h:708,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/internal/gtest-internal.h:40,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/gtest.h:58,
                 from /home/gvaillan/local/src/arrayfire/test/clamp.cpp:10:
/usr/include/c++/6/array:90:12: note:                 template<class _Tp, long unsigned int _Nm> struct std::array
     struct array
            ^~~~~
/home/gvaillan/local/src/arrayfire/test/clamp.cpp:32:5: error: ‘out’ was not declared in this scope
     out.host(&hout[0]);
     ^~~
/home/gvaillan/local/src/arrayfire/test/clamp.cpp:33:5: error: ‘in’ was not declared in this scope
     in.host(&hin[0]);
     ^~
/home/gvaillan/local/src/arrayfire/test/clamp.cpp: In member function ‘virtual void ClampTests_FloatArrayScalar_Test::TestBody()’:
/home/gvaillan/local/src/arrayfire/test/clamp.cpp:46:5: error: reference to ‘array’ is ambiguous
     array in = af::randu(num, f32);
     ^~~~~
In file included from /home/gvaillan/local/src/arrayfire/test/clamp.cpp:11:0:
/home/gvaillan/local/src/arrayfire/include/af/array.h:27:17: note: candidates are: class af::array
     class AFAPI array {
                 ^~~~~
In file included from /usr/include/c++/6/tuple:39:0,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/internal/gtest-port.h:708,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/internal/gtest-internal.h:40,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/gtest.h:58,
                 from /home/gvaillan/local/src/arrayfire/test/clamp.cpp:10:
/usr/include/c++/6/array:90:12: note:                 template<class _Tp, long unsigned int _Nm> struct std::array
     struct array
            ^~~~~
/home/gvaillan/local/src/arrayfire/test/clamp.cpp:47:5: error: reference to ‘array’ is ambiguous
     array lo = af::randu(num, f32)/10; // Ensure lo <= 0.1
     ^~~~~
In file included from /home/gvaillan/local/src/arrayfire/test/clamp.cpp:11:0:
/home/gvaillan/local/src/arrayfire/include/af/array.h:27:17: note: candidates are: class af::array
     class AFAPI array {
                 ^~~~~
In file included from /usr/include/c++/6/tuple:39:0,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/internal/gtest-port.h:708,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/internal/gtest-internal.h:40,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/gtest.h:58,
                 from /home/gvaillan/local/src/arrayfire/test/clamp.cpp:10:
/usr/include/c++/6/array:90:12: note:                 template<class _Tp, long unsigned int _Nm> struct std::array
     struct array
            ^~~~~
/home/gvaillan/local/src/arrayfire/test/clamp.cpp:51:5: error: reference to ‘array’ is ambiguous
     array out = clamp(in, lo, hi);
     ^~~~~
In file included from /home/gvaillan/local/src/arrayfire/test/clamp.cpp:11:0:
/home/gvaillan/local/src/arrayfire/include/af/array.h:27:17: note: candidates are: class af::array
     class AFAPI array {
                 ^~~~~
In file included from /usr/include/c++/6/tuple:39:0,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/internal/gtest-port.h:708,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/internal/gtest-internal.h:40,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/gtest.h:58,
                 from /home/gvaillan/local/src/arrayfire/test/clamp.cpp:10:
/usr/include/c++/6/array:90:12: note:                 template<class _Tp, long unsigned int _Nm> struct std::array
     struct array
            ^~~~~
/home/gvaillan/local/src/arrayfire/test/clamp.cpp:53:5: error: ‘out’ was not declared in this scope
     out.host(&hout[0]);
     ^~~
/home/gvaillan/local/src/arrayfire/test/clamp.cpp:54:5: error: ‘in’ was not declared in this scope
     in.host(&hin[0]);
     ^~
/home/gvaillan/local/src/arrayfire/test/clamp.cpp:55:5: error: ‘lo’ was not declared in this scope
     lo.host(&hlo[0]);
     ^~
/home/gvaillan/local/src/arrayfire/test/clamp.cpp: In member function ‘virtual void ClampTests_FloatScalarArray_Test::TestBody()’:
/home/gvaillan/local/src/arrayfire/test/clamp.cpp:66:5: error: reference to ‘array’ is ambiguous
     array in = af::randu(num, f32);
     ^~~~~
In file included from /home/gvaillan/local/src/arrayfire/test/clamp.cpp:11:0:
/home/gvaillan/local/src/arrayfire/include/af/array.h:27:17: note: candidates are: class af::array
     class AFAPI array {
                 ^~~~~
In file included from /usr/include/c++/6/tuple:39:0,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/internal/gtest-port.h:708,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/internal/gtest-internal.h:40,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/gtest.h:58,
                 from /home/gvaillan/local/src/arrayfire/test/clamp.cpp:10:
/usr/include/c++/6/array:90:12: note:                 template<class _Tp, long unsigned int _Nm> struct std::array
     struct array
            ^~~~~
/home/gvaillan/local/src/arrayfire/test/clamp.cpp:68:5: error: reference to ‘array’ is ambiguous
     array hi = 1.0 - af::randu(num, f32)/10; // Ensure hi >= 0.9
     ^~~~~
In file included from /home/gvaillan/local/src/arrayfire/test/clamp.cpp:11:0:
/home/gvaillan/local/src/arrayfire/include/af/array.h:27:17: note: candidates are: class af::array
     class AFAPI array {
                 ^~~~~
In file included from /usr/include/c++/6/tuple:39:0,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/internal/gtest-port.h:708,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/internal/gtest-internal.h:40,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/gtest.h:58,
                 from /home/gvaillan/local/src/arrayfire/test/clamp.cpp:10:
/usr/include/c++/6/array:90:12: note:                 template<class _Tp, long unsigned int _Nm> struct std::array
     struct array
            ^~~~~
/home/gvaillan/local/src/arrayfire/test/clamp.cpp:71:5: error: reference to ‘array’ is ambiguous
     array out = clamp(in, lo, hi);
     ^~~~~
In file included from /home/gvaillan/local/src/arrayfire/test/clamp.cpp:11:0:
/home/gvaillan/local/src/arrayfire/include/af/array.h:27:17: note: candidates are: class af::array
     class AFAPI array {
                 ^~~~~
In file included from /usr/include/c++/6/tuple:39:0,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/internal/gtest-port.h:708,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/internal/gtest-internal.h:40,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/gtest.h:58,
                 from /home/gvaillan/local/src/arrayfire/test/clamp.cpp:10:
/usr/include/c++/6/array:90:12: note:                 template<class _Tp, long unsigned int _Nm> struct std::array
     struct array
            ^~~~~
/home/gvaillan/local/src/arrayfire/test/clamp.cpp:73:5: error: ‘out’ was not declared in this scope
     out.host(&hout[0]);
     ^~~
/home/gvaillan/local/src/arrayfire/test/clamp.cpp:74:5: error: ‘in’ was not declared in this scope
     in.host(&hin[0]);
     ^~
/home/gvaillan/local/src/arrayfire/test/clamp.cpp:75:5: error: ‘hi’ was not declared in this scope
     hi.host(&hhi[0]);
     ^~
/home/gvaillan/local/src/arrayfire/test/clamp.cpp: In member function ‘virtual void ClampTests_FloatScalarScalar_Test::TestBody()’:
/home/gvaillan/local/src/arrayfire/test/clamp.cpp:86:5: error: reference to ‘array’ is ambiguous
     array in = af::randu(num, f32);
     ^~~~~
In file included from /home/gvaillan/local/src/arrayfire/test/clamp.cpp:11:0:
/home/gvaillan/local/src/arrayfire/include/af/array.h:27:17: note: candidates are: class af::array
     class AFAPI array {
                 ^~~~~
In file included from /usr/include/c++/6/tuple:39:0,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/internal/gtest-port.h:708,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/internal/gtest-internal.h:40,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/gtest.h:58,
                 from /home/gvaillan/local/src/arrayfire/test/clamp.cpp:10:
/usr/include/c++/6/array:90:12: note:                 template<class _Tp, long unsigned int _Nm> struct std::array
     struct array
            ^~~~~
/home/gvaillan/local/src/arrayfire/test/clamp.cpp:91:5: error: reference to ‘array’ is ambiguous
     array out = clamp(in, lo, hi);
     ^~~~~
In file included from /home/gvaillan/local/src/arrayfire/test/clamp.cpp:11:0:
/home/gvaillan/local/src/arrayfire/include/af/array.h:27:17: note: candidates are: class af::array
     class AFAPI array {
                 ^~~~~
In file included from /usr/include/c++/6/tuple:39:0,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/internal/gtest-port.h:708,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/internal/gtest-internal.h:40,
                 from /home/gvaillan/local/src/arrayfire/test/../test/gtest/include/gtest/gtest.h:58,
                 from /home/gvaillan/local/src/arrayfire/test/clamp.cpp:10:
/usr/include/c++/6/array:90:12: note:                 template<class _Tp, long unsigned int _Nm> struct std::array
     struct array
            ^~~~~
/home/gvaillan/local/src/arrayfire/test/clamp.cpp:93:5: error: ‘out’ was not declared in this scope
     out.host(&hout[0]);
     ^~~
/home/gvaillan/local/src/arrayfire/test/clamp.cpp:94:5: error: ‘in’ was not declared in this scope
     in.host(&hin[0]);
     ^~

```

--- a/src/backend/cuda/kernel/homography.hpp
+++ b/src/backend/cuda/kernel/homography.hpp
@@ -65,64 +65,54 @@ __device__ void JacobiSVD(int m, int n)
     int tid_y = threadIdx.y;
     //int gid_y = blockIdx.y * blockDim.y + tid_y;
 
-    __shared__ T acc1[256];
-    __shared__ T acc2[256];
+    __shared__ T s_acc1[256];
+    __shared__ T s_acc2[256];
 
-    __shared__ T d[16*9];
+    __shared__ T s_d[16*9];
 
     T* s_V = (T*)sh;
     T* s_S = (T*)sh + 16*81;
 
-    // Copy first 80 elements
-    for (int i = 0; i <= 4; i++) {
-        T t = s_S[tid_y*81 + tid_x+i*bsz_x];
-        acc1[tid_y*bsz_x + tid_x] += t*t;
-    }
-    if (tid_x < 8)
-        acc1[tid_y*16 + tid_x] += acc1[tid_y*16 + tid_x+8];
-    __syncthreads();
-    if (tid_x < 4)
-        acc1[tid_y*16 + tid_x] += acc1[tid_y*16 + tid_x+4];
-    __syncthreads();
-    if (tid_x < 2)
-        acc1[tid_y*16 + tid_x] += acc1[tid_y*16 + tid_x+2];
-    __syncthreads();
-    if (tid_x < 1) {
-        // Copy last element
-        T t = s_S[tid_y*bsz_x + tid_x+80];
-        acc1[tid_y*16 + tid_x] += acc1[tid_y*16 + tid_x+1] + t*t;
-    }
-    __syncthreads();
 
-    if (tid_x < n)
-        d[tid_y*9 + tid_x] = acc1[tid_y*bsz_x + tid_x];
+    int doff = tid_y *  n;
+    int soff = tid_y * 81;
 
-    // V is initialized as an identity matrix
-    for (int i = 0; i <= 4; i++) {
-        s_V[tid_y*81 + i*bsz_x + tid_x] = 0;
+    if (tid_x < n) {
+        T acc1 = 0;
+        for (int i = 0; i < m; i++) {
+            int stid = soff + tid_x * m + i;
+            T t = s_S[stid];
+            acc1 += t * t;
+            s_V[stid] = (tid_x == i) ? 1 : 0;
+        }
+        s_d[doff + tid_x] = acc1;
     }
-    __syncthreads();
-    if (tid_x < m)
-        s_V[tid_y*81 + tid_x*m + tid_x] = 1;
     __syncthreads();
 
     for (int it = 0; it < iterations; it++) {
-        bool converged = false;
-
         for (int i = 0; i < n-1; i++) {
             for (int j = i+1; j < n; j++) {
-                T* Si = s_S + tid_y*81 + i*m;
-                T* Sj = s_S + tid_y*81 + j*m;
+                T* Si = s_S + soff + i*m;
+                T* Sj = s_S + soff + j*m;
+
+                T* Vi = s_V + soff + i*n;
+                T* Vj = s_V + soff + j*n;
 
                 T p = (T)0;
                 for (int k = 0; k < m; k++)
                     p += Si[k]*Sj[k];
 
-                T c = 0, s = 0;
+                T di = s_d[doff + i];
+                T dj = s_d[doff + j];
+                __syncthreads();
 
-                bool cond = (abs(p) > EPS<T>::eps()*sqrt(d[tid_y*9 + i]*d[tid_y*9 + j]));
+                T c = 0, s = 0;
+                T t0 = 0, t1 = 0;
+                int cond = (fabs(p) > m*EPS<T>::eps()*sqrt(di * dj));
+                T a = 0, b = 0;
+
                 if (cond) {
-                    T y = d[tid_y*9 + i] - d[tid_y*9 + j];
+                    T y = di - dj;
                     T r = hypot(p*2, y);
                     T r2 = r*2;
                     if (y >= 0) {
@@ -133,61 +123,43 @@ __device__ void JacobiSVD(int m, int n)
                         s = sqrt((r - y) / r2);
                         c = p / (r2*s);
                     }
+
+                    for (int k = tid_x; k < m; k+=bsz_x) {
+                        t0 = c*Si[k] + s*Sj[k];
+                        t1 = c*Sj[k] - s*Si[k];
+                        Si[k] = t0;
+                        Sj[k] = t1;
+                        s_acc1[tid_y * bsz_x + k] = t0 * t0;
+                        s_acc2[tid_y * bsz_x + k] = t1 * t1;
+                    }
                 }
                 __syncthreads();
 
-                if (cond && tid_x < m) {
-                    T t0 = c*Si[tid_x] + s*Sj[tid_x];
-                    T t1 = c*Sj[tid_x] - s*Si[tid_x];
-                    Si[tid_x] = t0;
-                    Sj[tid_x] = t1;
-
-                    acc1[tid_y*16 + tid_x] = t0*t0;
-                    acc2[tid_y*16 + tid_x] = t1*t1;
+                if (cond) {
+                    a = 0;
+                    b = 0;
+                    for (int k = 0; k < m; k++) {
+                        a += s_acc1[tid_y * bsz_x + k];
+                        b += s_acc2[tid_y * bsz_x + k];
+                    }
+                    s_d[doff + i] = a;
+                    s_d[doff + j] = b;
                 }
                 __syncthreads();
 
-                if (cond && tid_x < 4) {
-                    acc1[tid_y*16 + tid_x] += acc1[tid_y*16 + tid_x+4];
-                    acc2[tid_y*16 + tid_x] += acc2[tid_y*16 + tid_x+4];
+                if (cond) {
+                    for (int l = tid_x; l < n; l += bsz_x) {
+                        T t0 = Vi[l] * c + Vj[l] * s;
+                        T t1 = Vj[l] * c - Vi[l] * s;
+
+                        Vi[l] = t0;
+                        Vj[l] = t1;
+                    }
                 }
                 __syncthreads();
-                if (cond && tid_x < 2) {
-                    acc1[tid_y*16 + tid_x] += acc1[tid_y*16 + tid_x+2];
-                    acc2[tid_y*16 + tid_x] += acc2[tid_y*16 + tid_x+2];
-                }
-                __syncthreads();
-                if (cond && tid_x < 1) {
-                    acc1[tid_y*16 + tid_x] += acc1[tid_y*16 + tid_x+1] + acc1[tid_y*16 + tid_x+8];
-                    acc2[tid_y*16 + tid_x] += acc2[tid_y*16 + tid_x+1] + acc2[tid_y*16 + tid_x+8];
-                }
-                __syncthreads();
-
-                if (cond && tid_x == 0) {
-                    d[tid_y*9 + i] = acc1[tid_y*16];
-                    d[tid_y*9 + j] = acc2[tid_y*16];
-                }
-                __syncthreads();
-
-                T* Vi = s_V + tid_y*81 + i*n;
-                T* Vj = s_V + tid_y*81 + j*n;
-
-                if (cond && tid_x < n) {
-                    T t0 = Vi[tid_x] * c + Vj[tid_x] * s;
-                    T t1 = Vj[tid_x] * c - Vi[tid_x] * s;
-
-                    Vi[tid_x] = t0;
-                    Vj[tid_x] = t1;
-                }
-                __syncthreads();
-
-                converged = true;
             }
-            if (!converged)
-                break;
         }
     }
-    __syncthreads();
 }
 
 __device__ bool computeMeanScale(
@@ -411,6 +383,7 @@ __global__ void computeEvalHomography(
             }
         }
     }
+    __syncthreads();
 
     if (htype == AF_HOMOGRAPHY_RANSAC) {
         // Find sample with most inliers
@@ -680,30 +653,14 @@ int computeH(
         memFree(totalInliers.ptr);
         memFree(median.ptr);
     } else if (htype == AF_HOMOGRAPHY_RANSAC) {
-        Param<unsigned> bestInliers, bestIdx;
-        for (int k = 0; k < 4; k++) {
-            bestInliers.dims[k] = bestIdx.dims[k] = 1;
-            bestInliers.strides[k] = bestIdx.strides[k] = 1;
-        }
-        bestInliers.ptr = memAlloc<unsigned>(1);
-        bestIdx.ptr = memAlloc<unsigned>(1);
-
-        kernel::ireduce<unsigned, af_max_t>(bestInliers, bestIdx.ptr, inliers, 0);
-
         unsigned blockIdx;
-        CUDA_CHECK(cudaMemcpyAsync(&blockIdx, bestIdx.ptr, sizeof(unsigned),
-                    cudaMemcpyDeviceToHost, cuda::getStream(cuda::getActiveDeviceId())));
-
+        inliersH = kernel::ireduce_all<unsigned, af_max_t>(&blockIdx, inliers);
         // Copies back index and number of inliers of best homography estimation
         CUDA_CHECK(cudaMemcpyAsync(&idxH, idx.ptr+blockIdx, sizeof(unsigned),
-                    cudaMemcpyDeviceToHost, cuda::getStream(cuda::getActiveDeviceId())));
-        CUDA_CHECK(cudaMemcpyAsync(&inliersH, bestInliers.ptr, sizeof(unsigned),
                     cudaMemcpyDeviceToHost, cuda::getStream(cuda::getActiveDeviceId())));
         CUDA_CHECK(cudaMemcpyAsync(bestH.ptr, H.ptr + idxH * 9, 9*sizeof(T),
                     cudaMemcpyDeviceToDevice, cuda::getStream(cuda::getActiveDeviceId())));
 
-        memFree(bestInliers.ptr);
-        memFree(bestIdx.ptr);
     }
 
     memFree(inliers.ptr);

--- a/src/backend/opencl/kernel/homography.cl
+++ b/src/backend/opencl/kernel/homography.cl
@@ -22,58 +22,57 @@ inline void jacobi_svd(__local T* l_V, __local T* l_S, __local T* l_d,
     int tid_y = get_local_id(1);
     int gid_y = get_global_id(1);
 
-    // Copy first 80 elements
-    T t = l_S[tid_y*81 + tid_x];
-    l_acc1[tid_y*bsz_x + tid_x] = t*t;
-    for (int i = 1; i <= 4; i++) {
-        T t = l_S[tid_y*81 + tid_x+i*bsz_x];
-        l_acc1[tid_y*bsz_x + tid_x] += t*t;
-    }
-    if (tid_x < 8)
-        l_acc1[tid_y*16 + tid_x] += l_acc1[tid_y*16 + tid_x+8];
-    barrier(CLK_LOCAL_MEM_FENCE);
-    if (tid_x < 4)
-        l_acc1[tid_y*16 + tid_x] += l_acc1[tid_y*16 + tid_x+4];
-    barrier(CLK_LOCAL_MEM_FENCE);
-    if (tid_x < 2)
-        l_acc1[tid_y*16 + tid_x] += l_acc1[tid_y*16 + tid_x+2];
-    barrier(CLK_LOCAL_MEM_FENCE);
-    if (tid_x < 1) {
-        // Copy last element
-        T t = l_S[tid_y*bsz_x + tid_x+80];
-        l_acc1[tid_y*16 + tid_x] += l_acc1[tid_y*16 + tid_x+1] + t*t;
+    int doff = tid_y *  n;
+    int soff = tid_y * 81;
+
+    if (tid_x < n) {
+        T acc1 = 0;
+        for (int i = 0; i < m; i++) {
+            int stid = soff + tid_x * m + i;
+            T t = l_S[stid];
+            acc1 += t * t;
+            l_V[stid] = (tid_x == i) ? 1 : 0;
+        }
+        l_d[doff + tid_x] = acc1;
     }
     barrier(CLK_LOCAL_MEM_FENCE);
 
-    if (tid_x < n)
-        l_d[tid_y*9 + tid_x] = l_acc1[tid_y*bsz_x + tid_x];
+#if defined(IS_CPU)
+    // All threads do the same work
+    // FIXME: Figure out why code below doesnt work
+    int tst = 0, toff = 1, tcond = tid_x == 0;
+#define BARRIER //nothing
+#else
+    // Split work across subgroup
+    int tst = tid_x, toff = bsz_x, tcond = 1;
+#define BARRIER barrier(CLK_LOCAL_MEM_FENCE)
+#endif
 
-    // V is initialized as an identity matrix
-    for (int i = 0; i <= 4; i++) {
-        l_V[tid_y*81 + i*bsz_x + tid_x] = 0;
-    }
-    barrier(CLK_LOCAL_MEM_FENCE);
-    if (tid_x < m)
-        l_V[tid_y*81 + tid_x*m + tid_x] = 1;
-    barrier(CLK_LOCAL_MEM_FENCE);
 
-    for (int it = 0; it < iterations; it++) {
-        int converged = 0;
-
+    for (int it = 0; tcond && it < iterations; it++) {
         for (int i = 0; i < n-1; i++) {
             for (int j = i+1; j < n; j++) {
-                __local T* Si = l_S + tid_y*81 + i*m;
-                __local T* Sj = l_S + tid_y*81 + j*m;
+                __local T* Si = l_S + soff + i*m;
+                __local T* Sj = l_S + soff + j*m;
+
+                __local T* Vi = l_V + soff + i*n;
+                __local T* Vj = l_V + soff + j*n;
 
                 T p = (T)0;
                 for (int k = 0; k < m; k++)
                     p += Si[k]*Sj[k];
 
-                T c = 0, s = 0;
+                T di = l_d[doff + i];
+                T dj = l_d[doff + j];
+                BARRIER;
 
-                int cond = (fabs(p) > EPS*sqrt(l_d[tid_y*9 + i]*l_d[tid_y*9 + j]));
+                T c = 0, s = 0;
+                T t0 = 0, t1 = 0;
+                int cond = (fabs(p) > m*EPS*sqrt(di * dj));
+                T a = 0, b = 0;
+
                 if (cond) {
-                    T y = l_d[tid_y*9 + i] - l_d[tid_y*9 + j];
+                    T y = di - dj;
                     T r = hypot(p*2, y);
                     T r2 = r*2;
                     if (y >= 0) {
@@ -85,59 +84,42 @@ inline void jacobi_svd(__local T* l_V, __local T* l_S, __local T* l_d,
                         c = p / (r2*s);
                     }
 
-                    if (tid_x < m) {
-                        T t0 = c*Si[tid_x] + s*Sj[tid_x];
-                        T t1 = c*Sj[tid_x] - s*Si[tid_x];
-                        Si[tid_x] = t0;
-                        Sj[tid_x] = t1;
-
-                        l_acc1[tid_y*16 + tid_x] = t0*t0;
-                        l_acc2[tid_y*16 + tid_x] = t1*t1;
+                    for (int k = tst; k < m; k+=toff) {
+                        t0 = c*Si[k] + s*Sj[k];
+                        t1 = c*Sj[k] - s*Si[k];
+                        Si[k] = t0;
+                        Sj[k] = t1;
+                        l_acc1[tid_y * bsz_x + k] = t0 * t0;
+                        l_acc2[tid_y * bsz_x + k] = t1 * t1;
                     }
                 }
-                barrier(CLK_LOCAL_MEM_FENCE);
+                BARRIER;
 
-                if (cond && tid_x < 4) {
-                    l_acc1[tid_y*16 + tid_x] += l_acc1[tid_y*16 + tid_x+4];
-                    l_acc2[tid_y*16 + tid_x] += l_acc2[tid_y*16 + tid_x+4];
+                if (cond) {
+                    a = 0;
+                    b = 0;
+                    for (int k = 0; k < m; k++) {
+                        a += l_acc1[tid_y * bsz_x + k];
+                        b += l_acc2[tid_y * bsz_x + k];
+                    }
+                    l_d[doff + i] = a;
+                    l_d[doff + j] = b;
                 }
-                barrier(CLK_LOCAL_MEM_FENCE);
-                if (cond && tid_x < 2) {
-                    l_acc1[tid_y*16 + tid_x] += l_acc1[tid_y*16 + tid_x+2];
-                    l_acc2[tid_y*16 + tid_x] += l_acc2[tid_y*16 + tid_x+2];
+                BARRIER;
+
+                if (cond) {
+                    for (int l = tst; l < n; l += toff) {
+                        T t0 = Vi[l] * c + Vj[l] * s;
+                        T t1 = Vj[l] * c - Vi[l] * s;
+
+                        Vi[l] = t0;
+                        Vj[l] = t1;
+                    }
                 }
-                barrier(CLK_LOCAL_MEM_FENCE);
-                if (cond && tid_x < 1) {
-                    l_acc1[tid_y*16 + tid_x] += l_acc1[tid_y*16 + tid_x+1] + l_acc1[tid_y*16 + tid_x+8];
-                    l_acc2[tid_y*16 + tid_x] += l_acc2[tid_y*16 + tid_x+1] + l_acc2[tid_y*16 + tid_x+8];
-                }
-                barrier(CLK_LOCAL_MEM_FENCE);
-
-                if (cond && tid_x == 0) {
-                    l_d[tid_y*9 + i] = l_acc1[tid_y*16];
-                    l_d[tid_y*9 + j] = l_acc2[tid_y*16];
-                }
-                barrier(CLK_LOCAL_MEM_FENCE);
-
-                __local T* Vi = l_V + tid_y*81 + i*n;
-                __local T* Vj = l_V + tid_y*81 + j*n;
-
-                if (cond && tid_x < n) {
-                    T t0 = Vi[tid_x] * c + Vj[tid_x] * s;
-                    T t1 = Vj[tid_x] * c - Vi[tid_x] * s;
-
-                    Vi[tid_x] = t0;
-                    Vj[tid_x] = t1;
-                }
-                barrier(CLK_LOCAL_MEM_FENCE);
-
-                converged = 1;
+                BARRIER;
             }
-            if (converged == 0)
-                break;
         }
     }
-    barrier(CLK_LOCAL_MEM_FENCE);
 }
 
 inline int compute_mean_scale(
@@ -194,7 +176,7 @@ inline int compute_mean_scale(
     *src_scale = sqrt(2.0f) / sqrt(src_var);
     *dst_scale = sqrt(2.0f) / sqrt(dst_var);
 
-    return !bad;
+    return bad;
 }
 
 #define LSPTR(Z, Y, X) (l_S[(Z) * 81 + (Y) * 9 + (X)])
@@ -212,19 +194,20 @@ __kernel void compute_homography(
 {
     unsigned i = get_global_id(1);
     unsigned tid_y = get_local_id(1);
+    unsigned tid_x = get_local_id(0);
 
     float x_src_mean, y_src_mean;
     float x_dst_mean, y_dst_mean;
     float src_scale, dst_scale;
     float src_pt_x[4], src_pt_y[4], dst_pt_x[4], dst_pt_y[4];
 
-    compute_mean_scale(&x_src_mean, &y_src_mean,
-                       &x_dst_mean, &y_dst_mean,
-                       &src_scale, &dst_scale,
-                       src_pt_x, src_pt_y,
-                       dst_pt_x, dst_pt_y,
-                       x_src, y_src, x_dst, y_dst,
-                       rnd, rInfo, i);
+    int bad = compute_mean_scale(&x_src_mean, &y_src_mean,
+                                 &x_dst_mean, &y_dst_mean,
+                                 &src_scale, &dst_scale,
+                                 src_pt_x, src_pt_y,
+                                 dst_pt_x, dst_pt_y,
+                                 x_src, y_src, x_dst, y_dst,
+                                 rnd, rInfo, i);
 
     __local T l_acc1[256];
     __local T l_acc2[256];
@@ -234,33 +217,33 @@ __kernel void compute_homography(
     __local T l_d[16*9];
 
     // Compute input matrix
-    for (unsigned j = get_local_id(0); j < 4; j+=get_local_size(0)) {
-        float srcx = (src_pt_x[j] - x_src_mean) * src_scale;
-        float srcy = (src_pt_y[j] - y_src_mean) * src_scale;
-        float dstx = (dst_pt_x[j] - x_dst_mean) * dst_scale;
-        float dsty = (dst_pt_y[j] - y_dst_mean) * dst_scale;
+    if (tid_x < 4) {
+        float srcx = (src_pt_x[tid_x] - x_src_mean) * src_scale;
+        float srcy = (src_pt_y[tid_x] - y_src_mean) * src_scale;
+        float dstx = (dst_pt_x[tid_x] - x_dst_mean) * dst_scale;
+        float dsty = (dst_pt_y[tid_x] - y_dst_mean) * dst_scale;
 
-        LSPTR(tid_y, 0, j*2) = 0.0f;
-        LSPTR(tid_y, 1, j*2) = 0.0f;
-        LSPTR(tid_y, 2, j*2) = 0.0f;
-        LSPTR(tid_y, 3, j*2) = -srcx;
-        LSPTR(tid_y, 4, j*2) = -srcy;
-        LSPTR(tid_y, 5, j*2) = -1.0f;
-        LSPTR(tid_y, 6, j*2) = dsty*srcx;
-        LSPTR(tid_y, 7, j*2) = dsty*srcy;
-        LSPTR(tid_y, 8, j*2) = dsty;
+        LSPTR(tid_y, 0, tid_x*2) = 0.0f;
+        LSPTR(tid_y, 1, tid_x*2) = 0.0f;
+        LSPTR(tid_y, 2, tid_x*2) = 0.0f;
+        LSPTR(tid_y, 3, tid_x*2) = -srcx;
+        LSPTR(tid_y, 4, tid_x*2) = -srcy;
+        LSPTR(tid_y, 5, tid_x*2) = -1.0f;
+        LSPTR(tid_y, 6, tid_x*2) = dsty*srcx;
+        LSPTR(tid_y, 7, tid_x*2) = dsty*srcy;
+        LSPTR(tid_y, 8, tid_x*2) = dsty;
 
-        LSPTR(tid_y, 0, j*2+1) = srcx;
-        LSPTR(tid_y, 1, j*2+1) = srcy;
-        LSPTR(tid_y, 2, j*2+1) = 1.0f;
-        LSPTR(tid_y, 3, j*2+1) = 0.0f;
-        LSPTR(tid_y, 4, j*2+1) = 0.0f;
-        LSPTR(tid_y, 5, j*2+1) = 0.0f;
-        LSPTR(tid_y, 6, j*2+1) = -dstx*srcx;
-        LSPTR(tid_y, 7, j*2+1) = -dstx*srcy;
-        LSPTR(tid_y, 8, j*2+1) = -dstx;
+        LSPTR(tid_y, 0, tid_x*2+1) = srcx;
+        LSPTR(tid_y, 1, tid_x*2+1) = srcy;
+        LSPTR(tid_y, 2, tid_x*2+1) = 1.0f;
+        LSPTR(tid_y, 3, tid_x*2+1) = 0.0f;
+        LSPTR(tid_y, 4, tid_x*2+1) = 0.0f;
+        LSPTR(tid_y, 5, tid_x*2+1) = 0.0f;
+        LSPTR(tid_y, 6, tid_x*2+1) = -dstx*srcx;
+        LSPTR(tid_y, 7, tid_x*2+1) = -dstx*srcy;
+        LSPTR(tid_y, 8, tid_x*2+1) = -dstx;
 
-        if (j == 4) {
+        if (tid_x == 4) {
             LSPTR(tid_y, 0, 8) = 0.0f;
             LSPTR(tid_y, 1, 8) = 0.0f;
             LSPTR(tid_y, 2, 8) = 0.0f;
@@ -276,28 +259,30 @@ __kernel void compute_homography(
 
     jacobi_svd(l_V, l_S, l_d, l_acc1, l_acc2, 9, 9);
 
-    T vH[9], H_tmp[9];
-    for (unsigned j = 0; j < 9; j++)
-        vH[j] = l_V[tid_y * 81 + 8 * 9 + j];
+    if (i < HInfo.dims[1] && tid_x == 0) {
+        T vH[9], H_tmp[9];
+        for (unsigned j = 0; j < 9; j++)
+            vH[j] = l_V[tid_y * 81 + 8 * 9 + j];
 
-    H_tmp[0] = src_scale*x_dst_mean*vH[6] + src_scale*vH[0]/dst_scale;
-    H_tmp[1] = src_scale*x_dst_mean*vH[7] + src_scale*vH[1]/dst_scale;
-    H_tmp[2] = x_dst_mean*(vH[8] - src_scale*y_src_mean*vH[7] - src_scale*x_src_mean*vH[6]) +
-                          (vH[2] - src_scale*y_src_mean*vH[1] - src_scale*x_src_mean*vH[0])/dst_scale;
+        H_tmp[0] = src_scale*x_dst_mean*vH[6] + src_scale*vH[0]/dst_scale;
+        H_tmp[1] = src_scale*x_dst_mean*vH[7] + src_scale*vH[1]/dst_scale;
+        H_tmp[2] = x_dst_mean*(vH[8] - src_scale*y_src_mean*vH[7] - src_scale*x_src_mean*vH[6]) +
+            (vH[2] - src_scale*y_src_mean*vH[1] - src_scale*x_src_mean*vH[0])/dst_scale;
 
-    H_tmp[3] = src_scale*y_dst_mean*vH[6] + src_scale*vH[3]/dst_scale;
-    H_tmp[4] = src_scale*y_dst_mean*vH[7] + src_scale*vH[4]/dst_scale;
-    H_tmp[5] = y_dst_mean*(vH[8] - src_scale*y_src_mean*vH[7] - src_scale*x_src_mean*vH[6]) +
-                          (vH[5] - src_scale*y_src_mean*vH[4] - src_scale*x_src_mean*vH[3])/dst_scale;
+        H_tmp[3] = src_scale*y_dst_mean*vH[6] + src_scale*vH[3]/dst_scale;
+        H_tmp[4] = src_scale*y_dst_mean*vH[7] + src_scale*vH[4]/dst_scale;
+        H_tmp[5] = y_dst_mean*(vH[8] - src_scale*y_src_mean*vH[7] - src_scale*x_src_mean*vH[6]) +
+            (vH[5] - src_scale*y_src_mean*vH[4] - src_scale*x_src_mean*vH[3])/dst_scale;
 
-    H_tmp[6] = src_scale*vH[6];
-    H_tmp[7] = src_scale*vH[7];
-    H_tmp[8] = vH[8] - src_scale*y_src_mean*vH[7] - src_scale*x_src_mean*vH[6];
+        H_tmp[6] = src_scale*vH[6];
+        H_tmp[7] = src_scale*vH[7];
+        H_tmp[8] = vH[8] - src_scale*y_src_mean*vH[7] - src_scale*x_src_mean*vH[6];
 
-    const unsigned Hidx = HInfo.dims[0] * i;
-    __global T* H_ptr = H + Hidx;
-    for (int h = 0; h < 9; h++)
-        H_ptr[h] = H_tmp[h];
+        const unsigned Hidx = HInfo.dims[0] * i;
+        __global T* H_ptr = H + Hidx;
+        for (int h = 0; h < 9; h++)
+            H_ptr[h] = bad ? 0 : H_tmp[h];
+    }
 }
 
 #undef APTR
@@ -364,6 +349,7 @@ __kernel void eval_homography(
         }
 #endif
     }
+    barrier(CLK_LOCAL_MEM_FENCE);
 
 #ifdef RANSAC
     unsigned bid_x = get_group_id(0);
@@ -379,8 +365,10 @@ __kernel void eval_homography(
         barrier(CLK_LOCAL_MEM_FENCE);
     }
 
-    inliers[bid_x] = l_inliers[0];
-    idx[bid_x]     = l_idx[0];
+    if (tid_x == 0) {
+        inliers[bid_x] = l_inliers[0];
+        idx[bid_x]     = l_idx[0];
+    }
 #endif
 }
 

--- a/src/backend/opencl/kernel/homography.hpp
+++ b/src/backend/opencl/kernel/homography.hpp
@@ -77,6 +77,10 @@ int computeH(
                 else if (htype == AF_HOMOGRAPHY_LMEDS)
                     options << " -D LMEDS";
 
+                if (getActiveDeviceType() == CL_DEVICE_TYPE_CPU) {
+                    options << " -D IS_CPU";
+                }
+
                 cl::Program prog;
                 buildProgram(prog, homography_cl, homography_cl_len, options.str());
                 hgProgs[device] = new Program(prog);
@@ -215,30 +219,14 @@ int computeH(
             getQueue().enqueueReadBuffer(*totalInliers.data, CL_TRUE, 0, sizeof(unsigned), &inliersH);
 
             bufferFree(totalInliers.data);
-        }
-        else if (htype == AF_HOMOGRAPHY_RANSAC) {
-            Param bestInliers, bestIdx;
-            bestInliers.info.offset = bestIdx.info.offset = 0;
-            for (int k = 0; k < 4; k++) {
-                bestInliers.info.dims[k] = bestIdx.info.dims[k] = 1;
-                bestInliers.info.strides[k] = bestIdx.info.strides[k] = 1;
-            }
-            bestInliers.data = bufferAlloc(sizeof(unsigned));
-            bestIdx.data = bufferAlloc(sizeof(unsigned));
-
-            kernel::ireduce<unsigned, af_max_t>(bestInliers, bestIdx.data, inliers, 0);
-
+        } else if (htype == AF_HOMOGRAPHY_RANSAC) {
             unsigned blockIdx;
-            getQueue().enqueueReadBuffer(*bestIdx.data, CL_TRUE, 0, sizeof(unsigned), &blockIdx);
+            inliersH = kernel::ireduce_all<unsigned, af_max_t>(&blockIdx, inliers);
 
             // Copies back index and number of inliers of best homography estimation
-            getQueue().enqueueReadBuffer(*idx.data, CL_TRUE, blockIdx*sizeof(unsigned), sizeof(unsigned), &idxH);
-            getQueue().enqueueReadBuffer(*bestInliers.data, CL_TRUE, 0, sizeof(unsigned), &inliersH);
-
+            getQueue().enqueueReadBuffer(*idx.data, CL_TRUE, blockIdx*sizeof(unsigned),
+                                         sizeof(unsigned), &idxH);
             getQueue().enqueueCopyBuffer(*H.data, *bestH.data, idxH*9*sizeof(T), 0, 9*sizeof(T));
-
-            bufferFree(bestInliers.data);
-            bufferFree(bestIdx.data);
         }
 
         bufferFree(inliers.data);

--- a/test/assign.cpp
+++ b/test/assign.cpp
@@ -115,7 +115,6 @@ void assignTest(string pTestFile, const vector<af_seq> *seqv)
     ASSERT_EQ(AF_SUCCESS, af_get_data_ptr((void*)outData, outArray));
 
     vector<outType> currGoldBar = tests[0];
-    using namespace std;
     size_t nElems        = currGoldBar.size();
     for (size_t elIter=0; elIter<nElems; ++elIter) {
         ASSERT_EQ(currGoldBar[elIter], outData[elIter])<< "at: " << elIter<< std::endl;
@@ -345,7 +344,6 @@ TYPED_TEST(ArrayAssign, Scalar4DCPP)
 TYPED_TEST(ArrayAssign, AssignRowCPP)
 {
     if (noDoubleTests<TypeParam>()) return;
-    using namespace std;
     using namespace af;
     int dimsize=10;
     vector<TypeParam> input(100, 1);
@@ -390,7 +388,6 @@ TYPED_TEST(ArrayAssign, AssignRowCPP)
 TYPED_TEST(ArrayAssign, AssignColumnCPP)
 {
     if (noDoubleTests<TypeParam>()) return;
-    using namespace std;
     using namespace af;
     int dimsize=10;
     vector<TypeParam> input(100, 1);
@@ -435,7 +432,6 @@ TYPED_TEST(ArrayAssign, AssignColumnCPP)
 TYPED_TEST(ArrayAssign, AssignSliceCPP)
 {
     if (noDoubleTests<TypeParam>()) return;
-    using namespace std;
     using namespace af;
     int dimsize=10;
     vector<TypeParam> input(1000, 1);

--- a/test/basic.cpp
+++ b/test/basic.cpp
@@ -13,7 +13,7 @@
 #include <vector>
 #include <testHelpers.hpp>
 
-using namespace std;
+using std::vector;
 
 TEST(BasicTests, constant1000x1000)
 {

--- a/test/binary.cpp
+++ b/test/binary.cpp
@@ -13,6 +13,7 @@
 #include <af/data.h>
 #include <testHelpers.hpp>
 
+// This makes the macros cleaner
 using namespace std;
 using std::abs;
 using namespace af;

--- a/test/clamp.cpp
+++ b/test/clamp.cpp
@@ -13,7 +13,6 @@
 #include <af/data.h>
 #include <testHelpers.hpp>
 
-using namespace std;
 using std::abs;
 using namespace af;
 
@@ -21,14 +20,14 @@ const int num = 10000;
 
 TEST(ClampTests, FloatArrayArray)
 {
-    array in = af::randu(num, f32);
-    array lo = af::randu(num, f32)/10;       // Ensure lo <= 0.1
-    array hi = 1.0 - af::randu(num, f32)/10; // Ensure hi >= 0.9
+    af::array in = af::randu(num, f32);
+    af::array lo = af::randu(num, f32)/10;       // Ensure lo <= 0.1
+    af::array hi = 1.0 - af::randu(num, f32)/10; // Ensure hi >= 0.9
     af::eval(lo, hi);
 
 
     std::vector<float> hout(num), hin(num), hlo(num), hhi(num);
-    array out = clamp(in, lo, hi);
+    af::array out = clamp(in, lo, hi);
     out.host(&hout[0]);
     in.host(&hin[0]);
     lo.host(&hlo[0]);
@@ -43,12 +42,12 @@ TEST(ClampTests, FloatArrayArray)
 
 TEST(ClampTests, FloatArrayScalar)
 {
-    array in = af::randu(num, f32);
-    array lo = af::randu(num, f32)/10; // Ensure lo <= 0.1
+    af::array in = af::randu(num, f32);
+    af::array lo = af::randu(num, f32)/10; // Ensure lo <= 0.1
     float hi = 0.9;
 
     std::vector<float> hout(num), hin(num), hlo(num);
-    array out = clamp(in, lo, hi);
+    af::array out = clamp(in, lo, hi);
 
     out.host(&hout[0]);
     in.host(&hin[0]);
@@ -63,12 +62,12 @@ TEST(ClampTests, FloatArrayScalar)
 
 TEST(ClampTests, FloatScalarArray)
 {
-    array in = af::randu(num, f32);
+    af::array in = af::randu(num, f32);
     float lo = 0.1;
-    array hi = 1.0 - af::randu(num, f32)/10; // Ensure hi >= 0.9
+    af::array hi = 1.0 - af::randu(num, f32)/10; // Ensure hi >= 0.9
 
     std::vector<float> hout(num), hin(num), hhi(num);
-    array out = clamp(in, lo, hi);
+    af::array out = clamp(in, lo, hi);
 
     out.host(&hout[0]);
     in.host(&hin[0]);
@@ -83,12 +82,12 @@ TEST(ClampTests, FloatScalarArray)
 
 TEST(ClampTests, FloatScalarScalar)
 {
-    array in = af::randu(num, f32);
+    af::array in = af::randu(num, f32);
     float lo = 0.1;
     float hi = 0.9;
 
     std::vector<float> hout(num), hin(num);
-    array out = clamp(in, lo, hi);
+    af::array out = clamp(in, lo, hi);
 
     out.host(&hout[0]);
     in.host(&hin[0]);

--- a/test/complex.cpp
+++ b/test/complex.cpp
@@ -13,7 +13,6 @@
 #include <af/data.h>
 #include <testHelpers.hpp>
 
-using namespace std;
 using namespace af;
 
 const int num = 10;

--- a/test/flat.cpp
+++ b/test/flat.cpp
@@ -13,7 +13,6 @@
 #include <af/data.h>
 #include <testHelpers.hpp>
 
-using namespace std;
 using namespace af;
 
 TEST(FlatTests, Test_flat_1D)

--- a/test/flip.cpp
+++ b/test/flip.cpp
@@ -14,7 +14,6 @@
 #include <af/index.h>
 #include <testHelpers.hpp>
 
-using namespace std;
 using namespace af;
 
 TEST(FlipTests, Test_flip_1D)

--- a/test/jit.cpp
+++ b/test/jit.cpp
@@ -13,7 +13,6 @@
 #include <af/data.h>
 #include <testHelpers.hpp>
 
-using namespace std;
 using namespace af;
 
 TEST(JIT, CPP_JIT_HASH)

--- a/test/math.cpp
+++ b/test/math.cpp
@@ -12,6 +12,7 @@
 #include <af/data.h>
 #include <testHelpers.hpp>
 
+// This makes the macros cleaner
 using namespace std;
 using namespace af;
 using std::abs;

--- a/test/ocl_ext_context.cpp
+++ b/test/ocl_ext_context.cpp
@@ -15,8 +15,7 @@
 
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Wdeprecated-declarations"
-
-using namespace std;
+using std::vector;
 
 inline void checkErr(cl_int err, const char * name) {
     if (err != CL_SUCCESS) {

--- a/test/stdev.cpp
+++ b/test/stdev.cpp
@@ -18,8 +18,9 @@
 #include <algorithm>
 #include <testHelpers.hpp>
 
-using namespace std;
 using namespace af;
+using std::string;
+using std::vector;
 
 template<typename T>
 class StandardDev : public ::testing::Test

